ray-operator/controllers/ray/common/service.go

```diff
@@ -40,25 +40,37 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
 		annotations = make(map[string]string)
 	}
 
+	serviceSpec := cluster.Spec.HeadGroupSpec.HeadServiceSpec
+
+	// For backward compatibility, use ServiceType if provided in HeadGroupSpec.
+	if cluster.Spec.HeadGroupSpec.ServiceType != "" {
+		serviceSpec.Type = cluster.Spec.HeadGroupSpec.ServiceType
+	}
+
+	// Use default labels if serviceSpec.Selector is not provided in HeadServiceSpec.
+	if serviceSpec.Selector == nil {
+		serviceSpec.Selector = labels
+	}
+
+	// Generate ports for service if Ports is not provided in HeadServiceSpec.
+	if serviceSpec.Ports == nil {
+		serviceSpec.Ports = []corev1.ServicePort{}
+		ports := getServicePorts(cluster)
+		defaultAppProtocol := DefaultServiceAppProtocol
+		for name, port := range ports {
+			svcPort := corev1.ServicePort{Name: name, Port: port, AppProtocol: &defaultAppProtocol}
+			serviceSpec.Ports = append(serviceSpec.Ports, svcPort)
+		}
+	}
+
 	service := &corev1.Service{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:        utils.GenerateServiceName(cluster.Name),
 			Namespace:   cluster.Namespace,
 			Labels:      labels,
 			Annotations: annotations,
 		},
-		Spec: corev1.ServiceSpec{
-			Selector: labels,
-			Ports:    []corev1.ServicePort{},
-			Type:     cluster.Spec.HeadGroupSpec.ServiceType,
-		},
-	}
-
-	ports := getServicePorts(cluster)
-	defaultAppProtocol := DefaultServiceAppProtocol
-	for name, port := range ports {
-		svcPort := corev1.ServicePort{Name: name, Port: port, AppProtocol: &defaultAppProtocol}
-		service.Spec.Ports = append(service.Spec.Ports, svcPort)
+		Spec: serviceSpec,
```

Comment:
`v1.ServiceSpec` is not enough for users. For example, users need to configure head service's annotations for some ingress controllers (#708). In addition, users also want to configure service's labels (#874, [slack thread](https://ray-distributed.slack.com/archives/C02GFQ82JPM/p1673647485229799)).

Reply:
Resolved offline. I'll update this PR to expose the entire head service (which in particular includes ObjectMeta, which includes labels and annotations.)